### DISCUSSION TITLE
Using Zodios API on Next app directory

### DISCUSSION THREAD

**Discussion by edgahan** [ecyrbe/zodios#492] [Q&A]
Hi all, 
I"m using Zodios with my NextJS project and it works well. On the server components I use the generated api and on the client side components I use Zodios Hooks (thin later on React Query). 

My question/observation is that the requests in my server components are not getting deduped. I looked at possibly caching (https://nextjs.org/docs/app/building-your-application/data-fetching/caching) however I'm struggling to get this to work as well. 

Has anyone progressed further on this?

**Comment by ecyrbe** [comment]
I know a user that needed this and used zodios v11 beta sucessfully for this :
install `@zodios/core` and `@zodios/fetch` both on version `11.0.0-beta.19`.
You'll import Zodios client from the fetch library. this version compatible with V10 api definition.
Note that future versions of zodios v11 beta will break api definitions. But if you stay on beta 19, (pin the dependency), you're good to go
Also zodios v11 requires typescript v5

**Comment by edgahan** [comment]
Thanks for the reply. I did try the above but unfortunately it didn't work. In case it is helpful to others, this is my setup and what I tried and what worked/didn't work.

First, for some context there is a NestJS backend and NextJS app directory frontend.
On the backend I create API docs and then I use https://github.com/astahmer/openapi-zod-client to generate an API in the Next app. 
 
 This basically works and I end up with something like the screenshot: 
 <img width="682" alt="Screenshot 2023-07-20 at 11 10 03" src="https://github.com/ecyrbe/zodios/assets/287499/03787ab8-f6f0-41ec-b6c5-b46e16bdd540">
 
 Inside of a `page.tsx` or a `layout.tsx` in the app directory I will use the API with something like:
```javascript
const accountUser: AccountUserType = await api.AccountUserController_findMe()
```

And inside of a client component I will have something like:
```javascript
const { mutate, error, isLoading } = apiHook.useClientActionsController_create(...)
```
  
Again this works, and it's typed and everything. 
  
My issue is if I have a `/nested/nested2/nested3` directory with a bunch of layouts, and the above snippet is present in each layout then the call happens multiple times on full page refresh. According to Next [docs](https://nextjs.org/docs/app/building-your-application/data-fetching/caching#per-request-caching) they should deduplicate if native `fetch` is used. 
  
In Zodios v10 by default it uses Axios, however I tried with the fetch plugin. 
I also tried upgrading to v11 as per above suggestion (I had to tweak a couple of things manually) but the TLDR is that the requests are still not being deduplicated.
  
I can get this to work on server components if I replace the above line with:
```javascript
const   accountUser = await getUser()
```

and 

```
import { cache } from "react"
export const getUser = cache(async () => {
    return generatedApi.AccountUserController_findMe()
})
```

And in theory with the `.hbs` template I could dynamically generate this for each endpoint.

**Reply by jebnascimento** [reply]
Hello guys, 

@edgahan could you share how you integrated zodios with the new app directory from NextJS 13? I followed the docs (https://www.zodios.org/docs/server/next), but it didn't work. 

I followed this example(https://github.com/ecyrbe/zodios-express/tree/main/examples/next), changing only what's in app directory, I have /api/[...zodios]/route.ts instead of /api/[...zodios].ts

**Reply by edgahan** [reply]
Sorry - I am not using Zodios on the backend like that, mostly just using it to generate the API from an Open API doc [generated by NestJS]

**Comment by jebnascimento** [comment]
> Sorry - I am not using Zodios on the backend like that, mostly just using it to generate the API from an Open API doc [generated by NestJS]
@edgahan thanks for your fast reply! 

@ecyrbe is there any barrier to use zodios with the new app directory? 

when I try to fetch any endpoint i've got a 405 and the following log: 

`[...zodios]/route.ts'. Export a named export for each HTTP method.`

I've tried to do that, but it didn't work


```
import { app } from "../../../server/routers/app";
export { app as GET, app as POST, app as PUT, app as DELETE};
```

**Reply by ecyrbe** [reply]
You can't use the server endpoints in app directory.
Because you need to export GET, POST, etc endpoints.
However you Can use Zodios client in app directory.

**Reply by jebnascimento** [reply]
Thanks man! I'll do that.
